emcache/client.py

```diff
@@ -769,6 +769,43 @@ async def gats_many(self, exptime: int, keys: Sequence[bytes], return_flags=Fals
 
         return results
 
+    async def auth(self, memcached_host_address: MemcachedHostAddress, username: bytes, password: bytes) -> None:
```

Comment:
Reading this [1] seems that the `auth` command is required for start sending commands to a recent open conection, so until this is not done the client will not be able to execute any command.

I have the feeling that we can not put the burden to have to do so explicitly every time, because among other stuff the client will not know if the connection that is being used behind the scenes did or did not sent the `auth` command.

Not sure how other memcache clients approach this, maybe we could research a bit here and see what is the prior art here, but having the feeling that one approach would be a one where the `auth` command is issued transparently behind the scenes every time that a new connection is created. If we go in that direction I would see more the `username` and `password` as part of the client constructor.

WDYT?


[1] https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L186

Reply:
exactly, let me transfer this to the initialization client with login and password

Reply:
@pfreixes, I can’t pass my own test for the second day. Why doesn't the client created first crash?

```python
    @pytest.mark.skipif(sys.platform == "darwin", reason="This server is not built with SASL support.")
    async def test_auth_with_errors(self):
        with pytest.raises(AuthenticationError):
            await create_client([MemcachedHostAddress("localhost", 11214)], timeout=2.0)
        with pytest.raises(AuthenticationError):
            await create_client([MemcachedHostAddress("localhost", 11214, username="wrong")], timeout=2.0)
        with pytest.raises(AuthenticationError):
            await create_client([MemcachedHostAddress("localhost", 11214, password="wrong")], timeout=2.0)

        with pytest.raises(AuthenticationNotSupported):
            await create_client([MemcachedHostAddress("localhost", 11211, "sad", "sad")], timeout=2.0)

        with pytest.raises(StorageCommandError):
            await create_client([MemcachedHostAddress("localhost", 11214, "wrong", "wrong")], timeout=2.0)
```

Reply:
> the password and login can be specified for several addresses in order to interact with node

IMO all hosts should use the same credentials. Too me this would be a requirement. We can check what other memcache clients do when implementing auth on top of a cluster of Memcached nodes, but having the feeling that they might be also asking for the same requirement.

> I can’t pass my own test for the second day. Why doesn't the client created first crash?

Is this something related to your changes, Im a bit lost here.

Reply:
IMO before going to implementation level, can we sketch the interfaces here? if we do that now there is a high chance that you don't need to redo your implementation later.

Reply:
or it would be better to set timeout=1.0 at the time of authentication?

Reply:
> Do you mind that now timeout initialization of the client will affect the same as commands?

Not sure how your implementation is handling the auth, but I would expect that since new connections will need to go through the auth step, the total timeout configured by the client for this specific first step will consider both commands, the auth + the command.

TBH, this would be expected to happen only once when a new connection is created.  Im ok on keep having for now just a single timeout, if users later on raise this as an issue we can always add a new timeout exclusively for the auth flow.

So, unless Im missing something here, I would go for keep it simple for now